lib/mongoid/association/embedded/batchable.rb

```diff
@@ -239,10 +239,11 @@ def normalize_docs(docs)
         #
         # @return [ String ] The atomic path.
         def path
-          @path ||= if _unscoped.empty?
+          return @path if @path
+          if _unscoped.empty?
             Mongoid::Atomic::Paths::Embedded::Many.position_without_document(_base, _association)
```

Comment:
The idea here is we don't want to set the cache if we are calculating the position without a document, since this method is often called before the association parent is set. This was causing the path to be set incorrectly for later on.